### ISSUE TITLE
Adding an option to disable file download on Jupyterlab

### ISSUE THREAD

**Issue by jmartinc89** [jupyterlab/jupyterlab#5274]
I think it would be useful to add an option in Jupyterlab to disable file download through the "Files" tab in the UI, maybe as a parameter at launch or using config files.

**Comment by schaffino** [comment]
+1 for this. We would like to allow disabling of upload and download from the jupyter UI and use a different managed service for this purpose.

**Comment by sulicny** [comment]
+1.  We would also appreciate this functionality.  We're hoping to use Jyptyer as an interface for researchers to analyze large private data sets, but restrict them from downloading the raw data the instances has access to.

**Comment by pelaprat** [comment]
@jmartinc89 Did you ever find a way to do this?

**Comment by Q4living** [comment]
Exhausted all the searches on google, really need this function ....

**Comment by hantswilliams** [comment]
Doing some investigation, might be able to modify source code, HTML template to remove it from the UI, per its git repo:  

`notebook/notebook/templates/notebook.html`

looks like this might be the part to comment out, then recompile: 

`<li class="dropdown-submenu" role="none"><a href="#" role="menuitem">{% trans %}Download as{% endtrans %}<span class="sr-only">{% trans %}Toggle Dropdown{% endtrans %}</span></a>
                            <ul id="download_menu" class="dropdown-menu">
                                {% for exporter in get_frontend_exporters() %}
                                <li id="download_{{ exporter.name }}">
                                    <a href="#">{{ exporter.display }}</a>
                                </li>
                                {% endfor %}
                            </ul>
                        </li>`

**Comment by vidartf** [comment]
Note: If this is intended as a security feature, there would also need to be code to change which server APIs are exposed, not just UI changes.

**Comment by isunix** [comment]
Anyone knows how to disable the download function for jupyterlab, not jupyter notebook? We really need this function to be removed in our development environment.

**Comment by jasongrout** [comment]
> Anyone knows how to disable the download function for jupyterlab, not jupyter notebook? We really need this function to be removed in our development environment.

Are you just wanting to disable the right-click menu option? Or do you want to actually disable downloading? The former can be straightforward. The latter will need notebook server changes and will likely be quite tricky to do.

**Comment by jasongrout** [comment]
To elaborate, as @vidartf points out, unless there are notebook server changes , the user can easily go to a special url and download any file.

**Comment by isunix** [comment]
@jasongrout  Disable the right-click menu option will be OK for me now? Can you please show me how to do so? I really need this function.

**Comment by isunix** [comment]
@jasongrout to add more info, I know how to disable the download function on the ui for jupyter notebook, but I can not find the way to do this on jupyterlab.

**Comment by dehiker** [comment]
@hantswilliams I comment out the similar part in
`anaconda3/pkgs/notebook-6.0.1-py37_0/lib/python3.7/site-packages/notebook/templates/notebook.html`
But I can see 'Download' button. Something wrong with what I did? Maybe how to "recompile" after commenting? Looking forward to your reply!

**Comment by h4gen** [comment]
Is there anything happening here? It seems to me that the many +1 show that this feature is needed. Not being able to prevent file downloads disqualifies jupyterlab for every use case where  data extraction e.g. from sensitive sources is prohibited.

**Comment by jasongrout** [comment]
As noted [above](https://github.com/jupyterlab/jupyterlab/issues/5274#issuecomment-566894040), this needs to be addressed in the `jupyter_server` and `notebook` projects to actually prevent file downloads. Just changing the JupyterLab frontend ux is just cosmetically changing things without actually preventing downloads.

**Comment by KLdivergence** [comment]
As @jasongrout point out, I managed to modified the download service by changing `base/handler.py` and `files/handler.py` in `notebook` and let `jupyter_server` import the modified `notebook` package. I did not do that, but I believe commenting `FilesHandler.get` in `files/handler.py` will be the easiest way to disable the download.

**Comment by DianeHu** [comment]
> > Anyone knows how to disable the download function for jupyterlab, not jupyter notebook? We really need this function to be removed in our development environment.
> 
> Are you just wanting to disable the right-click menu option? Or do you want to actually disable downloading? The former can be straightforward. The latter will need notebook server changes and will likely be quite tricky to do.

@jasongrout I've seen this thread mentioned multiple times and understand that removing the "download" button is just a cosmetic change, but this is also something I'm pursuing right now. You said disabling the menu option could be straightforward, but I also see https://github.com/jupyterlab/jupyterlab/issues/6115 state that this actually isn't doable. Is there an approach for removing this option cosmetically in the right-click context menu that is actually straightforward that I'm missing--can you elaborate?

**Comment by jasongrout** [comment]
This adds the download entry to the File menu: https://github.com/jupyterlab/jupyterlab/blob/5340e7161b4c5b67cd3e8184a018e142af86b22c/packages/docmanager-extension/src/index.ts#L730

This adds it to the command palette: https://github.com/jupyterlab/jupyterlab/blob/5340e7161b4c5b67cd3e8184a018e142af86b22c/packages/docmanager-extension/src/index.ts#L721

This is where it is added to the filebrowser context menu: https://github.com/jupyterlab/jupyterlab/blob/master/packages/filebrowser-extension/src/index.ts#L1098-L1102

A pattern we have in a few other places is instead of hard-coding adding those items to those various places, making a separate plugin in each of those extensions that adds the download option. Then you can omit those menu items by just disabling the appropriate plugin. To do this, someone needs to submit a PR moving the addition of those items to plugins. As a pattern, here is the plugin that adds the "Copy Shareable Link" menu item to the filebrowser context menu: https://github.com/jupyterlab/jupyterlab/blob/master/packages/filebrowser-extension/src/index.ts#L405-L455

**Comment by DianeHu** [comment]
@jasongrout ahh gotcha--thanks so much for the detailed explanation!

**Comment by cwchung85** [comment]
@DianeHu Thanks for the PRs! The ability to disable download is a sorely needed feature for our team. I do have a couple of follow up questions:
1. Do we have to rebuild from the source to have this download-disable feature? The preferred way for us is if we can simply install jupyterlab from repo like conda forge, and configure as admin to turn on the download-disable accordingly. 
2. We do have the capability to control the start up argument of jupyterlab, and therefore can provide additional runtime CLI parameter to enable this download-disable feature. Alternatively, I think there is something called overrides.json that might be useful for admin to enable/disable certain config. We can use either way to turn on the download-disable.

Once again, thanks very much!

**Comment by cwchung85** [comment]
@jasongrout   Further to this download-disable feature, I would be interested to work to disable the download at the server side, now that the GUI part have been addressed by Diane.  I read your previous comment that both notebook and jupyter_server module needs to be updated. In reading the source code of both, I got a couple of questions:
1. Apparently, jlab sends msg to jupyter_server and notebook for various tasks/requests. So what are being sent to jupyter_server and what are being sent to notebook? Could you point me to the right doc or source code?
2. If we assume download-disable feature is something that can be turned on or off by configuration, ideally the backend jupyter_server and notebook should also use the same configuration. Is it being the case today?
3. For backend related work, is this the right place for discussion, or should I post to a different subgroup? 
Thanks very much!

**Comment by jtpio** [comment]
> The preferred way for us is if we can simply install jupyterlab from repo like conda forge, and configure as admin to turn on the download-disable accordingly.

Yes that will be the case with the 3.1 release (or the next 3.1 alpha release), for example using page config parameters: https://jupyterlab.readthedocs.io/en/stable/user/extensions.html?#enabling-and-disabling-extensions

**Comment by DianeHu** [comment]
@cwchung85 npnp! For question 1, echoing what jtpio said above--once this goes into an official release you'll be able to just disable download by running ```jupyter labextension disable <id of the newly pulled out download extension>```, or editing the disabledExtensions field in your package.json.

**Comment by jtpio** [comment]
Actually `3.1.0a5` was released today: https://pypi.org/project/jupyterlab/3.1.0a5/

So you can give it a try. Thanks!

**Comment by cchung85** [comment]
> Actually `3.1.0a5` was released today: https://pypi.org/project/jupyterlab/3.1.0a5/
> 
> So you can give it a try. Thanks!

Nice! Let me give it a try!

**Comment by cchung85** [comment]
> @cwchung85 npnp! For question 1, echoing what jtpio said above--once this goes into an official release you'll be able to just disable download by running `jupyter labextension disable <id of the newly pulled out download extension>`, or editing the disabledExtensions field in your package.json.

Perfect! Thanks very much!

**Comment by cwchung85** [comment]
> Actually `3.1.0a5` was released today: https://pypi.org/project/jupyterlab/3.1.0a5/
> 
> So you can give it a try. Thanks!

I was able to set the page_config.json to disable a sample labextension jupyterlab-plotly.  Thanks very much!
So here  a couple of questions:
 
1. How to specify in this page_config.json to turn on the download-disable feature (that Diane's PRs)?
 
2. Jupyter --path gives the output below. I found that as admin, I can set up the page_config.json (owned by root) in the public directories to disable download, but the user can actually set up a similar page_config.json in his/her own directory to enable the download. I have tested and confirmed indeed the precedence does work that way, so the user can defeat the system set up. So is there a solution for that?  Thanks very much!!

$ jupyter --path
config:
    /home/dgreen/.jupyter   # owned by user dgreen who can set up page_config.json to enable download
    /opt/conda/envs/jupyterlab-310/etc/jupyter
    /usr/local/etc/jupyter
    /etc/jupyter
$ cat  /opt/conda/envs/jupyterlab-310/etc/jupyter/labconfig/page_config.json # to disable extension
{
  "disabledExtensions": {
    "jupyterlab-plotly": true
  }
}

**Comment by jtpio** [comment]
> How to specify in this page_config.json to turn on the download-disable feature (that Diane's PRs)?

Normally something like the following should do it:

```json
{
   "disabledExtensions": {
         "@jupyterlab/docmanager-extension:download": true
   }
}
```

**Comment by cwchung85** [comment]
> > How to specify in this page_config.json to turn on the download-disable feature (that Diane's PRs)?
> 
> Normally something like the following should do it:
> 
> ```json
> {
>    "disabledExtensions": {
>          "@jupyterlab/docmanager-extension:download": true
>    }
> }
> ```

Yes, this works beautifully!  Thanks very much for the tips! 
Kudos to @DianeHu and @jtpio !!  Well done!

So in summary for the benefits of other folks, to disable the download (from File dropdown and also context pop up):
A. Put this in a file labconfig/page_config.json in one of the directories returned from jupyter --path
{
   "disabledExtensions": {
         "@jupyterlab/docmanager-extension:download": true,
         "@jupyterlab/filebrowser-extension:download": true
   }
}

B. Admin can use JUPYTER_CONFIG_PATH to prepend an admin-provided system labconfig directory with page_config.json to have the highest precedence, such that these settings cannot be overridden by users.

**Comment by jasongrout** [comment]
> So in summary for the benefits of other folks, to disable the download (from File dropdown and also context pop up):

Thanks for making these instructions concise. And to be clear, this only disables the menu items, but does not actually disable downloading from the server.

**Comment by cwchung85** [comment]
@jasongrout 
Yes, understood this fix is for the UI only, but  that is sufficient to get the security off our back for a while :-)

Now for the other related fixes in jupyter_server and notebook modules, I have browsed the source code a bit, and I got a couple of questions:

1. So jupyterlab sends msg to jupyter_server and notebook for various tasks/requests. So what are being sent to jupyter_server and what are being sent to notebook? Could you point me to the right doc or source code?  (I think python code in a cell would be sent to notebook for execution, but beyond that, I don;t know much). 

2. So download disable can be enabled by configuration. Ideally the backend jupyter_server and notebook should also use the same configuration. Is it being the case today?  

3. For backend related work, is this the right place for discussion, or should I post to a different subgroup?

Thanks very much!

**Comment by jasongrout** [comment]
Backend work is a coordinated conversation with the Jupyter server project. Jlab uses the server rest API to get files: https://jupyter-server.readthedocs.io/en/stable/developers/rest-api.html

**Comment by kwon14** [comment]
여러분의 작업에 정말 감사드린다.
여러 답변들을 통해 적용을 해보고 있지만, 해결이 안되어 물어본다.
labconfig/page_config.json 는 찾을 수 없었다.
내 환경에서 jupyter --path를 하면, 아래와 같이 나온다.
------------------------------------------
# jupyter --path
config:
    /home/notebook/.jupyter
    /opt/conda/etc/jupyter
    /usr/local/etc/jupyter
    /etc/jupyter
data:
    /home/notebook/.local/share/jupyter
    /opt/conda/share/jupyter
    /usr/local/share/jupyter
    /usr/share/jupyter
runtime:
    /home/notebook/.local/share/jupyter/runtime
--------------------------------------------------
이 모든 경로를 찾아 보았지만, labconfig/page_config.json 는 찾을 수 없었다.

아래 내용을 어디에 넣을 수 있을지 물어본다.
부디 답변 부탁드린다.
감사합니다.
--------------------------------------------------
{
"disabledExtensions": {
"@jupyterlab/docmanager-extension:download": true,
"@jupyterlab/filebrowser-extension:download": true
}
}

**Comment by kwon14** [comment]
Thank you so much for your work.
I am trying to apply it through various answers, but I am asking because it cannot be resolved.
labconfig/page_config.json could not be found.
When I do jupyter --path in my environment, it looks like this:
------------------------------------------
# jupyter --path
config:
    /home/notebook/.jupyter
    /opt/conda/etc/jupyter
    /usr/local/etc/jupyter
    /etc/jupyter
data:
    /home/notebook/.local/share/jupyter
    /opt/conda/share/jupyter
    /usr/local/share/jupyter
    /usr/share/jupyter
runtime:
    /home/notebook/.local/share/jupyter/runtime
--------------------------------------------------
I searched all these paths, but couldn't find labconfig/page_config.json.

Ask where you can put the following.
Please answer me.
Thank you.
--------------------------------------------------
{
"disabledExtensions": {
"@jupyterlab/docmanager-extension:download": true,
"@jupyterlab/filebrowser-extension:download": true
}
}

**Comment by cwchung85** [comment]
I have did more testing on this download disable feature and so far so good.  I did ran into a case where download disable feature was observed to be NOT working and took me a bit of time to figure out. I thought I would share with the observations in the hope that it may save others some agony:
1. Version of jupyterlab from conda list and pip list: confirmed to be jupyterlab 3.1.0a5
2. The page_config.json was set up correctly, as witnessed by: jupyter labextension list showing both items have been disabled: 
Disabled extensions:
  @jupyterlab/docmanager-extension:download
  @jupyterlab/filebrowser-extension:download
So the config settings looks OK, right? However, when I clicked File, the Download entry was there! Also, when i did the right mouse click, both Download and Copy Download Link were still there. So what happened?

Upon troubleshooting, I found the following:
A. I did use pip install jupyterlab==3.1.0a5 and it installed jlab correctly
B. Somewhere along the line, I also did conda install jupyterlab-git
C. After that, conda list and pip list continued to show jupyterlab == 3.1.0a5.
D.. However, jupyter --version shows jupyter lab 3.0.14  !!

So what happened was: 
E. jupyterlab-git somehow did conda install jupyterlab 3.0.14 (oblivious to the fact that 3.1.0a5 was present)
F. jupyter somehow did recognize this 3.0.14 version, not the 3.1.0a5 version.
G. Jupyter-lab actually ran the version 3.0.14 which does not support download-disable
H. All the while jupyter labextension list still happily reported the disabling of these two downloads (which perhaps made sense, as it simply reported the config without knowing if the underlying software could do anything with that).

I am not sure which components perhaps needs improvement, but my workaround is not to install this jupyterlab-git for now. (I did conda install several other labextnsion such as jupyterlab-latex, and they worked OK and download disable worked.). Thx,

**Comment by DianeHu** [comment]
@cwchung85 for this problem, maybe in whatever build system you use, you could specify the version of jupyterlab required as being >=3.1.0a5? That way, older versions aren't valid. I also would try to stick with using only conda install, not pip install, as much as possible (basically, just stick with one or the other where you can). Sometimes the two don't play nicely together.

**Comment by cwchung85** [comment]
>> @cwchung85 thanks for noting this for future users! For this problem, maybe in whatever build system you use, you could specify the version of jupyterlab required as being >=3.1.0a5? That way, older versions aren't valid. I also would try to stick with using only conda install, not pip install, as much as possible (basically, just stick with one or the other where you can). Sometimes the two don't play nicely together.


Thanks for the suggestion...
In this case, jupyterlab 3.1.0a5 was installed by pip, whereas jupyterlab-git by conda. I think it is this difference that causes something not working. Jupyterlab-git evidently cannot find 3.1.0a5 as this is not available through conda channel. So >=3.1.0a5 would not work. I have verified in my testing, conda complained that nothing provides this version. 

I think once jupyterlab 3.1 versions becomes available in conda channel, this issue would go away.

**Comment by cwchung85** [comment]
You need to create a new file, as in all likelikhood such file wont; exist already.

Create the file in any of the following should work:

/home/notebook/.jupyter/labconfig/page_config.json
/opt/conda/etc/jupyter/labconfig/page_config.json
/usr/local/etc/jupyter/labconfig/page_config.json
/etc/jupyter/labconfig/page_config.json
Good luck!
    On Wednesday, April 28, 2021, 02:01:35 AM PDT, kwon14 ***@***.***> wrote:  
 
 


Thank you so much for your work.
I am trying to apply it through various answers, but I am asking because it cannot be resolved.
labconfig/page_config.json could not be found.
When I do jupyter --path in my environment, it looks like this:

jupyter --path

config:
/home/notebook/.jupyter
/opt/conda/etc/jupyter
/usr/local/etc/jupyter
/etc/jupyter
data:
/home/notebook/.local/share/jupyter
/opt/conda/share/jupyter
/usr/local/share/jupyter
/usr/share/jupyter
runtime:
/home/notebook/.local/share/jupyter/runtime

I searched all these paths, but couldn't find labconfig/page_config.json.

Ask where you can put the following.
Please answer me.
Thank you.

{
"disabledExtensions": {
***@***.***/docmanager-extension:download": true,
***@***.***/filebrowser-extension:download": true
}
}

**Comment by DianeHu** [comment]
> You need to create a new file, as in all likelikhood such file wont; exist already.

You can also put the disabledExtensions part into your package.json, it'll work the same way.

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/jupyterlab-config-issue-in-handling-read-only-config-dir-user-cannot-save-settings-such-as-dark-theme/9302/2

**Comment by TonicZhang** [comment]
I have disabled the download extension and removed the "launch classic notebook" in help menu. 
But user could just type browser url like "https://ip:port/location/tree" to launch a classic notebook, and then download freely…… anyone knows how to close the classic notebook server or disable download for classic notebook (not jupyterlab)

**Comment by jtpio** [comment]
> anyone knows how to close the classic notebook server or disable download for classic notebook (not jupyterlab)

With JupyterLab 4 and newer, the classic notebook is not installed anymore by default and should not be accessible from JupyterLab. That is if you install JupyterLab alone, for example with `pip install jupyterlab`.
